File: Graphs/MinimumJumpsToReachEndViaPrimeTeleportation.go
// LeetCode 3629 - Minimum Jumps to Reach End via Prime Teleportation
// Time Complexity: O(n log log M + n * d) | Space Complexity: O(M + n)
package main

const factorLimit = 1_000_001

var primeFactors = [factorLimit][]int{}

func init() {
	for factor := 2; factor < factorLimit; factor++ {
		if len(primeFactors[factor]) == 0 {
			for multiple := factor; multiple < factorLimit; multiple += factor {
				primeFactors[multiple] = append(primeFactors[multiple], factor)
			}
		}
	}
}

func minJumps(nums []int) int {
	n := len(nums)
	groups := map[int][]int{}

	for index, value := range nums {
		for _, factor := range primeFactors[value] {
			groups[factor] = append(groups[factor], index)
		}
	}

	visited := make([]bool, n)
	visited[0] = true
	queue := []int{0}
	jumps := 0

	for len(queue) > 0 {
		nextQueue := make([]int, 0)
		for _, index := range queue {
			if index == n-1 {
				return jumps
			}

			if index+1 < n && !visited[index+1] {
				visited[index+1] = true
				nextQueue = append(nextQueue, index+1)
			}
			if index-1 >= 0 && !visited[index-1] {
				visited[index-1] = true
				nextQueue = append(nextQueue, index-1)
			}

			nextIndices, ok := groups[nums[index]]
			if !ok {
				continue
			}

			for _, nextIndex := range nextIndices {
				if !visited[nextIndex] {
					visited[nextIndex] = true
					nextQueue = append(nextQueue, nextIndex)
				}
			}
			groups[nums[index]] = groups[nums[index]][:0]
		}
		queue = nextQueue
		jumps++
	}

	return -1
}

File: Graphs/MinimumJumpsToReachEndViaPrimeTeleportation.py
# LeetCode 3629 - Minimum Jumps to Reach End via Prime Teleportation
# Time Complexity: O(n log log M + n * d) | Space Complexity: O(M + n)
from collections import defaultdict, deque
from typing import Deque, Dict, List


LIMIT = 1_000_001
FACTORS = [[] for _ in range(LIMIT)]

for factor in range(2, LIMIT):
    if not FACTORS[factor]:
        for multiple in range(factor, LIMIT, factor):
            FACTORS[multiple].append(factor)


class Solution:
    def minJumps(self, nums: List[int]) -> int:
        groups: Dict[int, List[int]] = defaultdict(list)
        for index, value in enumerate(nums):
            for factor in FACTORS[value]:
                groups[factor].append(index)

        n = len(nums)
        visited = [False] * n
        visited[0] = True
        queue: Deque[int] = deque([0])
        jumps = 0

        while queue:
            for _ in range(len(queue)):
                index = queue.popleft()
                if index == n - 1:
                    return jumps

                if index + 1 < n and not visited[index + 1]:
                    visited[index + 1] = True
                    queue.append(index + 1)
                if index - 1 >= 0 and not visited[index - 1]:
                    visited[index - 1] = True
                    queue.append(index - 1)

                next_indices = groups.get(nums[index])
                if next_indices is None:
                    continue

                for next_index in next_indices:
                    if not visited[next_index]:
                        visited[next_index] = True
                        queue.append(next_index)

                next_indices.clear()

        return -1
